File: android/src/main/java/jp/manse/DAIManager.java
package jp.manse;

import android.content.Context;
import android.util.Log;
import android.view.ViewGroup;

import com.brightcove.player.model.DeliveryType;
import com.brightcove.player.model.Video;
import com.brightcove.player.view.BrightcoveExoPlayerVideoView;
import com.google.ads.interactivemedia.v3.api.AdErrorEvent;
import com.google.ads.interactivemedia.v3.api.AdEvent;
import com.google.ads.interactivemedia.v3.api.AdsLoader;
import com.google.ads.interactivemedia.v3.api.ImaSdkFactory;
import com.google.ads.interactivemedia.v3.api.StreamDisplayContainer;
import com.google.ads.interactivemedia.v3.api.StreamRequest;
import com.google.ads.interactivemedia.v3.api.player.VideoProgressUpdate;
import com.google.ads.interactivemedia.v3.api.player.VideoStreamPlayer;

import java.util.HashMap;
import java.util.List;

import jp.manse.util.Configuration;

public class DAIManager implements AdEvent.AdEventListener, AdErrorEvent.AdErrorListener {
    private BrightcoveExoPlayerVideoView videoPlayer;
    private ImaSdkFactory sdkFactory = ImaSdkFactory.getInstance();
    private AdsLoader adsLoader;
    private StreamDisplayContainer displayContainer;

    public DAIManager(Context context, ViewGroup adUiContainer, BrightcoveExoPlayerVideoView videoPlayer) {
        this.videoPlayer = videoPlayer;
        this.displayContainer = sdkFactory.createStreamDisplayContainer();
        displayContainer.setVideoStreamPlayer(videoStreamPlayer);
        displayContainer.setAdContainer(adUiContainer);
        adsLoader = sdkFactory.createAdsLoader(context, sdkFactory.createImaSdkSettings(), displayContainer);
    }

    @Override
    public void onAdError(AdErrorEvent adErrorEvent) {
        Log.d("DAIManager", "Event: " + adErrorEvent.toString());
    }

    @Override
    public void onAdEvent(AdEvent adEvent) {
        Log.d("DAIManager", "Error: " + adEvent.toString());
    }

    private VideoStreamPlayer videoStreamPlayer = new VideoStreamPlayer() {
        @Override
        public void loadUrl(String url, List<HashMap<String, String>> list) {
            if (url != null) {
                Video video = Video.createVideo(url, DeliveryType.HLS);
                videoPlayer.clear();
                videoPlayer.add(video);
                videoPlayer.start();
            }
        }

        @Override
        public int getVolume() {
            return 1;
        }

        @Override
        public void addCallback(VideoStreamPlayerCallback videoStreamPlayerCallback) {

        }

        @Override
        public void removeCallback(VideoStreamPlayerCallback videoStreamPlayerCallback) {

        }

        @Override
        public void onAdBreakStarted() {

        }

        @Override
        public void onAdBreakEnded() {

        }

        @Override
        public void onAdPeriodStarted() {

        }

        @Override
        public void onAdPeriodEnded() {

        }

        @Override
        public void seek(long l) {

        }

        @Override
        public VideoProgressUpdate getContentProgress() {
            return new VideoProgressUpdate(new Long(videoPlayer.getCurrentPosition()), new Long(videoPlayer.get(0).getDuration()));
        }
    };

    public void requestAndPlayAds() {
        adsLoader.addAdErrorListener(this);
        adsLoader.requestStream(buildStreamRequest());
    }

    private StreamRequest buildStreamRequest() {
        return sdkFactory.createLiveStreamRequest(Configuration.DAI_ASSET_KEY, null);
    }

}

File: android/src/main/java/jp/manse/DAIManager.kt
package jp.manse

import android.content.Context
import android.util.Log
import android.view.ViewGroup
import com.brightcove.player.model.DeliveryType
import com.brightcove.player.model.Video
import com.brightcove.player.view.BrightcoveExoPlayerVideoView
import com.google.ads.interactivemedia.v3.api.AdErrorEvent
import com.google.ads.interactivemedia.v3.api.AdEvent
import com.google.ads.interactivemedia.v3.api.AdsLoader
import com.google.ads.interactivemedia.v3.api.ImaSdkFactory
import com.google.ads.interactivemedia.v3.api.StreamDisplayContainer
import com.google.ads.interactivemedia.v3.api.StreamRequest
import jp.manse.util.Configuration
import com.google.ads.interactivemedia.v3.api.player.VideoStreamPlayer
import com.google.ads.interactivemedia.v3.api.player.VideoProgressUpdate


class DAIManager(private val context: Context,
                 private val adUiContainer: ViewGroup,
                 private val videoPlayer: BrightcoveExoPlayerVideoView
) : AdEvent.AdEventListener, AdErrorEvent.AdErrorListener {

    private val sdkFactory: ImaSdkFactory = ImaSdkFactory.getInstance()
    private lateinit var adsLoader: AdsLoader
    private var displayContainer: StreamDisplayContainer

    init {
        displayContainer = sdkFactory.createStreamDisplayContainer()
        createAdsLoader()
    }

    private fun createAdsLoader() {
        displayContainer = sdkFactory.createStreamDisplayContainer()
        val videoStreamPlayer = createVideoStreamPlayer()

        displayContainer.videoStreamPlayer = videoStreamPlayer
        displayContainer.adContainer = adUiContainer
        adsLoader = sdkFactory.createAdsLoader(context, sdkFactory.createImaSdkSettings(), displayContainer)
    }

    fun requestAndPlayAds() {
        adsLoader.addAdErrorListener(this)
        adsLoader.requestStream(buildStreamRequest())
    }

    private fun createVideoStreamPlayer() = object : VideoStreamPlayer {
            override fun onAdBreakEnded() {
            }

            override fun onAdBreakStarted() {
            }

            override fun seek(p0: Long) {
            }

            override fun getVolume(): Int {
                return 100
            }

            override fun removeCallback(p0: VideoStreamPlayer.VideoStreamPlayerCallback?) {
            }

            override fun onAdPeriodStarted() {
            }

            override fun loadUrl(url: String?, p1: MutableList<HashMap<String, String>>?) {
                Log.v("DAIMAN", ""+ url)
                if (url != null) {
                    val video = Video.createVideo(url, DeliveryType.HLS)
                    videoPlayer.clear()
                    videoPlayer.add(video)
                    videoPlayer.start()
                }
            }

            override fun addCallback(p0: VideoStreamPlayer.VideoStreamPlayerCallback?) {
            }

            override fun getContentProgress(): VideoProgressUpdate {
                return if (videoPlayer.duration <= 0) {
                    VideoProgressUpdate.VIDEO_TIME_NOT_READY
                } else {
                    VideoProgressUpdate(videoPlayer.currentPosition.toLong(),
                        videoPlayer.get(0).duration.toLong())
                }
            }

            override fun onAdPeriodEnded() {
            }

    }

    private fun buildStreamRequest(): StreamRequest {
        Log.v("DAIMAN", ""+ Configuration.DAI_ASSET_KEY)
        return sdkFactory.createLiveStreamRequest(Configuration.DAI_ASSET_KEY, null)
    }

    override fun onAdEvent(event: AdEvent?) {
        Log.d("DAIManager", "Event: " + event!!.type)
    }

    override fun onAdError(event: AdErrorEvent?) {
        Log.d("DAIManager", "Error: " + event!!.error.message)
    }
}
